ui/raidboss/data/05-shb/raid/e8s.js

```diff
@@ -317,30 +317,201 @@
+              // Both melee, so go on priority
+              if (dps1Priority > dps2Priority) {
+                // Swap, they're out of order
+                let tmp = data.rampant.chain.chains.dps1;
+                data.rampant.chain.chains.dps1 = data.rampant.chain.chains.dps2;
+                data.rampant.chain.chains.dps2 = tmp;
+              }
+            } else if (dps1Priority >= 0 || dps2Priority >= 0) {
+              if (dps2Priority >= 0) {
+                // Swap, they're out of order
+                let tmp = data.rampant.chain.chains.dps1;
+                data.rampant.chain.chains.dps1 = data.rampant.chain.chains.dps2;
+                data.rampant.chain.chains.dps2 = tmp;
+              }
+            } else {
+              // Both ranged, so go on priority
+              dps1Priority = rangedPriority.indexOf(dps1Role);
+              dps2Priority = rangedPriority.indexOf(dps2Role);
+              if (dps1Priority > dps2Priority) {
+                // Swap, they're out of order
+                let tmp = data.rampant.chain.chains.dps1;
+                data.rampant.chain.chains.dps1 = data.rampant.chain.chains.dps2;
+                data.rampant.chain.chains.dps2 = tmp;
+              }
+            }
+            // All positions based on tank tethers
+            let tankTether1 = (
+              data.rampant.chain.chains.tank.target === data.rampant.chain.chains.healer.source ? 'NE' : (
+                data.rampant.chain.chains.tank.target === data.rampant.chain.chains.dps1.source ? 'SW' : 'SE'
+              )
+            );
+            let tankTether2 = (
+              data.rampant.chain.chains.tankReverse.source === data.rampant.chain.chains.healer.target ? 'NE' : (
+                data.rampant.chain.chains.tankReverse.source === data.rampant.chain.chains.dps1.target ? 'SW' : 'SE'
+              )
+            );
+            if ((tankTether1 === 'NE' && tankTether2 === 'SW') ||
+                (tankTether1 === 'SW' && tankTether2 === 'NE')) {
+              // If the tank is tethered to NE and SW (makes a box)
+              // Tank swaps with SW to make bowtie
+              data.rampant.chain.chainSwapTank = 'SW';
+              data.rampant.chain.chainSwapDps = data.rampant.chain.chains.dps1.source;
+            } else if ((tankTether1 === 'NE' && tankTether2 === 'SE') ||
+                (tankTether1 === 'SE' && tankTether2 === 'NE')) {
+              // If the tank is tethered to NE and SE (makes an hourglass)
+              // Tank swaps with SE to make bowtie
+              data.rampant.chain.chainSwapTank = 'SE';
+              data.rampant.chain.chainSwapDps = data.rampant.chain.chains.dps2.source;
+            }
+          }
+          return data.rampant.chain.total===4;
+        }
+      },
+      infoText: function(data, matches) {
+        if (data.options.cactbote8sLightRampantStrat === undefined || data.options.cactbote8sLightRampantStrat === 'none') {
```

Comment:
From a 2- minute look, no. This info text should display only if there's no strategy defined, and in that event, the beginning of the condition *should* prevent the entire trigger activating more than once. (We've already established that we're confident there's exactly 1 tether that any chained player is the source for.)

(That being said, I'm not sure why we're checking both `undefined` and `none`, rather than doing something like `if (!data.options.cactbote8sLightRampantStrat)`.)

Also, maybe instead of stuffing all this into the condition we could do it as a `preRun`?

Reply:
> (That being said, I'm not sure why we're checking both `undefined` and `none`, rather than doing something like `if (!data.options.cactbote8sLightRampantStrat)`.)

`undefined` was a 'just in case that option hasn't been set' check, since I'm not sure exactly how Cactbot handles initializing those values if there's already a config present. The first version of this code block was a switch/case with the default being to return the 'chain on you' object, but that was ugly from a code readability standpoint.

> Also, maybe instead of stuffing all this into the condition we could do it as a preRun?

I don't know how feasible it would be to do this in a `preRun`. We need to do this logic regardless but we don't want to run the trigger fully unless we're at 4 triggers, and there's no way to abort a trigger after the `condition` that I've seen thus far.

I wish there was a way to move logic like this out to its own file, honestly.

Reply:
I'm probably missing something obvious, but in that case, I feel like the logic could be split off into the pre-run after incrementing the chain counter. It looks like the bulk of the logic there doesn't actually run anyway unless the condition would be met, so I'm not sure what the loss would be. (Obviously this has already been written as it is, but if it's "just" a matter of pulling lines 350-409 into the pre-run, I feel like that would be worthwhile.)

Good point on the config being present though, I forgot that this isn't guaranteed to be falsy if there's no strategy selected.

Reply:
The trigger execution process is essentially this:
```
if (condition returns true) {
    execute preRun
    execute infoText
    execute run
}
```
Therefore, we'd still need to store off the info into data, then increment the chain counter, and then in preRun/infoText/run/whatever we'd need to early out for cases where it isn't this strat. Either way it's irrelevant with the latest commit where the logic has been moved into response (which replaces infoText).